### PR TITLE
Fix file URLs

### PR DESCRIPTION
**Pull request by dwalluck**

### CODE REVIEW

File: cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java [hunk 1]
```diff
@@ -448,10 +449,10 @@ private void createSession( File target, File settings )
             {
                 // No, this is not a typo. If current default is null, supply new local and global.
                 // This function passes in settings to make it easier to test.
-                this.settings = settings = new File( System.getProperty( "user.home" ), ".m2/settings.xml" );
+                this.settings = settings = new File( System.getProperty( "user.home" ), ".m2" + File.separator + "settings.xml" );
```
Comment (dwalluck) [inline cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java]: This is obviously a fix for file separator not URL.

File: cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java [hunk 2]
```diff
@@ -465,7 +466,7 @@ private void createSession( File target, File settings )
 
             if ( ar != null)
             {
-                ar.setUrl( "file://" + req.getLocalRepositoryPath() );
+                ar.setUrl( req.getLocalRepositoryPath().toPath().toUri().toString() );
```
Comment (dwalluck) [inline cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java]: I assume that `File.toURI()` is the same as `Path.toUri()`, but just to be safe I converted it to a `Path` first. There is something int he javadoc about `File.toURL()` not being safe, in any case.

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java [hunk 3]
```diff
@@ -94,10 +94,12 @@
                 for ( final String script : scripts )
                 {
                     File found;
-                    if ( script.startsWith( "http" ) || script.startsWith( "file" ))
+                    logger.info( "Processing groovy script {}", script );
+                    if ( script.startsWith( "http:" ) || script.startsWith( "file:" ) )
```
Comment (dwalluck) [inline core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: Missing `:` here to mark the protocol.

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java [hunk 4]
```diff
@@ -94,10 +94,12 @@
...
                     {
-                        logger.info( "Attempting to read URL {} ", script );
-                        found = fileIO.resolveURL( new URL( script ) );
+                        URL url = new URL( script );
+                        logger.info( "Attempting to read URL {}", url );
+                        found = fileIO.resolveURL( url );
```
Comment (dwalluck) [inline core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: This is the line that does not like the `groovyScripts=file://` URLs which gives `UnknownHostException: C` on Windows later when it does `url.openStream()` in `fileIO.resolveURL()`.

Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: I don't see why this is needed?

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 3 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: I don't see why this is needed?

File: integration-test/src/it/circular-dependencies-test-second/test.properties [hunk 5]
```diff
@@ -17,5 +17,5 @@
 restURL=http://127.0.0.1:8089
 versionIncrementalSuffix=redhat
 dependencySource=rest
-groovyScripts=file://${basedir}/src/it/circular-dependencies-test-second/Sample.groovy
```
Comment (rnc) [inline integration-test/src/it/circular-dependencies-test-second/test.properties]: The `file://` as I thought we discussed previously is an acceptable standard so I don't see why this needs changing?

File: integration-test/src/it/groovy-manipulator-file/verify.groovy [hunk 6]
```diff
@@ -30,7 +30,7 @@ assert !jsonString.contains ("resolved:")
 def buildLog = new File( basedir, 'build.log' )
 def message = 0
 buildLog.eachLine {
-   if (it.contains( "groovyScripts file:///")) {
+   if (it.contains( "groovyScripts file:")) {
```
Comment (rnc) [inline integration-test/src/it/groovy-manipulator-file/verify.groovy]: Why this change?

File: integration-test/src/it/groovy-manipulator-file/test.properties [hunk 7]
```diff
@@ -16,4 +16,4 @@
 
 
 # The colon needs to be escaped to prevent the invoker plugin from misinterpreting it
-groovyScripts=file://${basedir}/src/it/setup/depMgmt1/Sample.groovy
```
Comment (rnc) [inline integration-test/src/it/groovy-manipulator-file/test.properties]: As per above

File: integration-test/src/it/remote-dependency-management-embedded-props-strict/test.properties [hunk 8]
```diff
@@ -21,4 +21,4 @@ dependencyManagement=org.commonjava.maven.ext.integration-test\:depMgmt2\:1.0.0-
 version.suffix=redhat-1
 overrideTransitive=false
 strictAlignment=true
-groovyScripts=file://${basedir}/src/it/remote-dependency-management-embedded-props-strict/inlineProp.groovy
```
Comment (rnc) [inline integration-test/src/it/remote-dependency-management-embedded-props-strict/test.properties]: As per above

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 3 above)
Comment (dwalluck) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: If you want to check if it's a URL, then the URL must start with `scheme + ':'`, or otherwise strings such as `"file.txt"` will match, which is unintended, and it could even match before the the final `else` case.

It also matches what it used elsewhere in the code:

https://github.com/release-engineering/pom-manipulation-ext/blob/8606c8b023b8a60333e41495d3ad38e466254491/core/src/main/java/org/commonjava/maven/ext/core/impl/RepoAndReportingRemovalManipulator.java#L264

https://github.com/release-engineering/pom-manipulation-ext/blob/8606c8b023b8a60333e41495d3ad38e466254491/io/src/main/java/org/commonjava/maven/ext/io/resolver/GalleyInfrastructure.java#L238

https://github.com/release-engineering/pom-manipulation-ext/blob/8606c8b023b8a60333e41495d3ad38e466254491/io/src/main/java/org/commonjava/maven/ext/io/resolver/GalleyInfrastructure.java#L266

https://github.com/release-engineering/pom-manipulation-ext/blob/8606c8b023b8a60333e41495d3ad38e466254491/io/src/main/java/org/commonjava/maven/ext/io/resolver/MavenLocationExpander.java#L115

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 4 above)
Comment (dwalluck) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: I was trying to see if there was a difference between the given input URL which is `script` and the url after it has been parsed which is `url`. This was for debugging, so `logger.debug` could have been used.

File: integration-test/src/it/circular-dependencies-test-second/test.properties (hunk 5 above)
Comment (dwalluck) [reply integration-test/src/it/circular-dependencies-test-second/test.properties]: URLs beginning with `file://` are definitely incorrect, but I can revert the changes to the properties files and try to fix up the invalid URL in the code instead. I can add a test case to show the failure.

Note that such a URL parses, so in that sense it is _valid_, but it's _semantically_ incorrect in that it expects an `authority` (hostname) to come after two slashes, which is true in the case of a network URL, but a file URL should have no `authority`

File: integration-test/src/it/groovy-manipulator-file/verify.groovy (hunk 6 above)
Comment (dwalluck) [reply integration-test/src/it/groovy-manipulator-file/verify.groovy]: Because if `${basedir}` does not begin with a `/` then `file://${basedir}` won't have three slashes and this check will fail, so it should be sufficient to check for `file:` which is what seems to be used everywhere else in the code anyway.

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 3 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: Ok

File: cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java (hunk 1 above)
Comment (rnc) [reply cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java]: I've changed this code so I think it might be good to rebase to master please

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 3 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: Ok sure, I guess being consistent is sensible.

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 4 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: As this can convert the original `script` value I think it would be better kept as it was.

File: integration-test/src/it/groovy-manipulator-file/verify.groovy (hunk 6 above)
Comment (rnc) [reply integration-test/src/it/groovy-manipulator-file/verify.groovy]: Ok

File: cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java (hunk 2 above)
Comment (rnc) [reply cli/src/main/java/org/commonjava/maven/ext/cli/Cli.java]: As per above, I think this needs rebasing.

File: integration-test/src/it/circular-dependencies-test-second/test.properties (hunk 5 above)
Comment (rnc) [reply integration-test/src/it/circular-dependencies-test-second/test.properties]: Lets discuss offline.

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java (hunk 4 above)
Comment (dwalluck) [reply core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: There's no code that changed here. I just moved it to two separate lines so I could log the difference between the values. It looks like `URL.toString()` at the very least was printing only one slash in some cases, which is strange. In any case, I can revert it.

File: integration-test/src/it/circular-dependencies-test-second/test.properties (hunk 5 above)
Comment (dwalluck) [reply integration-test/src/it/circular-dependencies-test-second/test.properties]: All I know is that `file:` with no slashes appeared to work on all OSes via the CI. With two or three slashes, there will need to be code to fix up the URI on Windows.

File: core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java [hunk 9]
```diff
@@ -69,6 +71,17 @@
 
     public abstract int getExecutionIndex();
 
+    String fixFileURL( final String script )
```
Comment (rnc) [inline core/src/main/java/org/commonjava/maven/ext/core/impl/BaseGroovyManipulator.java]: This means this method will be exposed to even the Gradle Manipulator tooling which is not ideal - given its internal only. Can we discuss offline please.

**Comment by dwalluck** [comment]
Obsoleted by #717
